File: packages/agroclima/climate-notifications/resources/android/ClimateNotificationFunctions.kt
package com.agroclima.plugins.climatenotifications

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.work.Constraints
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.NetworkType
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.nativephp.mobile.bridge.BridgeFunction
import com.nativephp.mobile.bridge.BridgeResponse
import com.nativephp.mobile.utils.NativeActionCoordinator
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.util.Locale
import java.util.concurrent.TimeUnit

object ClimateNotificationFunctions {
    private const val PERMISSION_EVENT = "App\\Events\\ClimateNotificationPermissionResult"

    class CheckPermission(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            activity.runOnUiThread {
                NotificationPermissionCoordinator.install(activity).check(
                    parameters["event"] as? String ?: PERMISSION_EVENT,
                    parameters["id"] as? String,
                )
            }

            return BridgeResponse.success(emptyMap())
        }
    }

    class RequestPermission(private val activity: FragmentActivity) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            activity.runOnUiThread {
                NotificationPermissionCoordinator.install(activity).request(
                    PermissionRequest(
                        event = parameters["event"] as? String ?: PERMISSION_EVENT,
                        id = parameters["id"] as? String,
                    ),
                )
            }

            return BridgeResponse.success(emptyMap())
        }
    }

    class SyncSchedule(private val context: Context) : BridgeFunction {
        override fun execute(parameters: Map<String, Any>): Map<String, Any> {
            ClimateAlertSchedule.sync(context, JSONObject(parameters).toString())

            return BridgeResponse.success(emptyMap())
        }
    }
}

private data class PermissionRequest(val event: String, val id: String?)

private class NotificationPermissionCoordinator : Fragment() {
    private val pendingRequests = mutableListOf<PermissionRequest>()
    private var permissionRequestInFlight = false

    private val notificationPermissionLauncher =
        registerForActivityResult(ActivityResultContracts.RequestPermission()) {
            permissionRequestInFlight = false
            val allowed = notificationsAllowed(requireContext())
            pendingRequests.toList().forEach { request ->
                dispatch(request.event, request.id, allowed)
            }
            pendingRequests.clear()
        }

    companion object {
        private const val FRAGMENT_TAG = "agroclima.notification-permission"

        fun install(activity: FragmentActivity): NotificationPermissionCoordinator {
            val manager = activity.supportFragmentManager
            val existing = manager.findFragmentByTag(FRAGMENT_TAG)

            if (existing is NotificationPermissionCoordinator) {
                return existing
            }

            return NotificationPermissionCoordinator().also { fragment ->
                manager.beginTransaction()
                    .add(fragment, FRAGMENT_TAG)
                    // A tab switch can arrive after Android saved the
                    // activity state. This coordinator is headless, so
                    // allowing state loss avoids killing the app while the
                    // next screen is being mounted; it will be found or
                    // installed again on the next request.
                    .commitNowAllowingStateLoss()
            }
        }
    }

    fun check(event: String, id: String?) {
        dispatch(event, id, notificationsAllowed(requireContext()))
    }

    fun request(request: PermissionRequest) {
        val context = requireContext()

        if (notificationsAllowed(context) || Build.VERSION.SDK_INT < 33) {
            dispatch(request.event, request.id, notificationsAllowed(context))
            return
        }

        pendingRequests.add(request)
        if (permissionRequestInFlight) {
            return
        }

        permissionRequestInFlight = true
        notificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
    }

    private fun dispatch(event: String, id: String?, allowed: Boolean) {
        val payload = JSONObject().put("granted", allowed)
        id?.let { payload.put("id", it) }
        NativeActionCoordinator.dispatchEvent(requireActivity(), event, payload.toString())
    }

    private fun notificationsAllowed(context: Context): Boolean {
        val runtimePermissionGranted = Build.VERSION.SDK_INT < 33 ||
            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) ==
            PackageManager.PERMISSION_GRANTED
        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val channelEnabled = Build.VERSION.SDK_INT < 26 || manager
            .getNotificationChannel("climate_alerts")
            ?.importance != NotificationManager.IMPORTANCE_NONE

        return runtimePermissionGranted && channelEnabled && NotificationManagerCompat.from(context).areNotificationsEnabled()
    }
}

private object ClimateAlertSchedule {
    private const val PREFERENCES = "agroclima_climate_notifications"
    private const val SCHEDULE_KEY = "schedule"
    private const val ALERT_STATES_KEY = "alert_states"
    private const val WORK_NAME = "agroclima-climate-alert-refresh"
    private const val CHANNEL_ID = "climate_alerts"
    private const val CHANNEL_NAME = "Alertas climáticas"
    private const val REFRESH_INTERVAL_HOURS = 1L

    fun sync(context: Context, scheduleJson: String) {
        val preferences = context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
        val schedule = JSONObject(scheduleJson)
        val locations = schedule.optJSONArray("locations") ?: JSONArray()
        val alerts = flattenAlerts(locations)
        val states = reconcileStates(preferences.getString(ALERT_STATES_KEY, "{}"), alerts)

        preferences.edit()
            .putString(SCHEDULE_KEY, schedule.toString())
            .putString(ALERT_STATES_KEY, states.toString())
            .apply()

        if (alerts.length() == 0 || !notificationsAllowed(context)) {
            WorkManager.getInstance(context).cancelUniqueWork(WORK_NAME)
            return
        }

        createNotificationChannel(context)
        val request = PeriodicWorkRequestBuilder<ClimateAlertWorker>(
            REFRESH_INTERVAL_HOURS,
            TimeUnit.HOURS,
        )
            .setConstraints(
                Constraints.Builder()
                    .setRequiredNetworkType(NetworkType.CONNECTED)
                    .build(),
            )
            .build()

        WorkManager.getInstance(context).enqueueUniquePeriodicWork(
            WORK_NAME,
            ExistingPeriodicWorkPolicy.UPDATE,
            request,
        )
    }

    fun cancel(context: Context) {
        WorkManager.getInstance(context).cancelUniqueWork(WORK_NAME)
    }

    fun notificationAllowed(context: Context): Boolean = notificationsAllowed(context)

    fun schedule(context: Context): JSONObject? = context
        .getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
        .getString(SCHEDULE_KEY, null)
        ?.let(::JSONObject)

    fun alertStates(context: Context): JSONObject = JSONObject(
        context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
            .getString(ALERT_STATES_KEY, "{}").orEmpty(),
    )

    fun saveAlertStates(context: Context, states: JSONObject) {
        context.getSharedPreferences(PREFERENCES, Context.MODE_PRIVATE)
            .edit()
            .putString(ALERT_STATES_KEY, filterStates(context, states).toString())
            .apply()
    }

    fun isActive(context: Context, alertId: String, signature: String): Boolean {
        val locations = schedule(context)?.optJSONArray("locations") ?: return false

        for (locationIndex in 0 until locations.length()) {
            val alerts = locations.optJSONObject(locationIndex)?.optJSONArray("alerts") ?: continue
            for (alertIndex in 0 until alerts.length()) {
                val alert = alerts.optJSONObject(alertIndex) ?: continue
                if (alert.optString("id") == alertId && alert.optString("signature") == signature) {
                    return true
                }
            }
        }

        return false
    }

    private fun filterStates(context: Context, states: JSONObject): JSONObject {
        val alerts = flattenAlerts(schedule(context)?.optJSONArray("locations") ?: JSONArray())
        val activeStates = JSONObject()

        for (index in 0 until alerts.length()) {
            val alert = alerts.optJSONObject(index) ?: continue
            val id = alert.optString("id")
            val state = states.optJSONObject(id) ?: continue
            if (state.optString("signature") == alert.optString("signature")) {
                activeStates.put(id, state)
            }
        }

        return activeStates
    }

    private fun flattenAlerts(locations: JSONArray): JSONArray {
        val alerts = JSONArray()

        for (index in 0 until locations.length()) {
            val location = locations.optJSONObject(index) ?: continue
            val locationAlerts = location.optJSONArray("alerts") ?: continue

            for (alertIndex in 0 until locationAlerts.length()) {
                val alert = locationAlerts.optJSONObject(alertIndex) ?: continue
                alerts.put(alert.put("locationId", location.optString("id")))
            }
        }

        return alerts
    }

    private fun reconcileStates(previousJson: String?, alerts: JSONArray): JSONObject {
        val previous = runCatching { JSONObject(previousJson ?: "{}") }.getOrDefault(JSONObject())
        val next = JSONObject()

        for (index in 0 until alerts.length()) {
            val alert = alerts.optJSONObject(index) ?: continue
            val id = alert.optString("id")
            val signature = alert.optString("signature")
            val initialState = alert.optString("lastState").takeUnless { it == "null" }
            val initialTimestamp = alert.optLong("lastEvaluatedAt", 0L)
            val saved = previous.optJSONObject(id)

            val state = if (saved == null || saved.optString("signature") != signature) {
                statePayload(signature, initialState, initialTimestamp)
            } else if (initialTimestamp > saved.optLong("evaluatedAt", 0L)) {
                statePayload(signature, initialState, initialTimestamp)
            } else {
                saved
            }

            next.put(id, state)
        }

        return next
    }

    private fun statePayload(signature: String, state: String?, timestamp: Long): JSONObject = JSONObject()
        .put("signature", signature)
        .put("state", state)
        .put("evaluatedAt", timestamp)

    private fun notificationsAllowed(context: Context): Boolean {
        val permissionGranted = Build.VERSION.SDK_INT < 33 ||
            ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) ==
            PackageManager.PERMISSION_GRANTED
        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val channelEnabled = Build.VERSION.SDK_INT < 26 || manager
            .getNotificationChannel(CHANNEL_ID)
            ?.importance != NotificationManager.IMPORTANCE_NONE

        return permissionGranted && channelEnabled && NotificationManagerCompat.from(context).areNotificationsEnabled()
    }

    private fun createNotificationChannel(context: Context) {
        if (Build.VERSION.SDK_INT < 26) {
            return
        }

        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        manager.createNotificationChannel(
            NotificationChannel(CHANNEL_ID, CHANNEL_NAME, NotificationManager.IMPORTANCE_DEFAULT),
        )
    }
}

class ClimateAlertWorker(
    context: Context,
    parameters: WorkerParameters,
) : Worker(context, parameters) {
    override fun doWork(): Result {
        if (!ClimateAlertSchedule.notificationAllowed(applicationContext)) {
            ClimateAlertSchedule.cancel(applicationContext)
            return Result.success()
        }

        val schedule = ClimateAlertSchedule.schedule(applicationContext) ?: return Result.success()
        val locations = schedule.optJSONArray("locations") ?: return Result.success()
        val forecastUrl = secureForecastUrl(schedule.optString("forecastUrl")) ?: return Result.failure()
        val states = ClimateAlertSchedule.alertStates(applicationContext)
        var hadRequestFailure = false

        for (index in 0 until locations.length()) {
            if (isStopped) {
                return Result.success()
            }

            val location = locations.optJSONObject(index) ?: continue
            val current = fetchCurrentWeather(
                forecastUrl,
                location.optDouble("latitude", Double.NaN),
                location.optDouble("longitude", Double.NaN),
            )

            if (current == null) {
                hadRequestFailure = true
                continue
            }

            val alerts = location.optJSONArray("alerts") ?: continue
            val now = System.currentTimeMillis()

            for (alertIndex in 0 until alerts.length()) {
                if (isStopped) {
                    return Result.success()
                }

                val alert = alerts.optJSONObject(alertIndex) ?: continue
                val id = alert.optString("id")
                val signature = alert.optString("signature")
                val metric = alert.optString("metric")
                val value = current[metric]
                val state = alertState(value, alert.optString("operator"), alert.optDouble("threshold", Double.NaN))
                val saved = states.optJSONObject(id)
                val previousState = if (saved == null || saved.optString("signature") != signature) {
                    alert.optString("lastState").takeUnless { it == "null" }
                } else {
                    saved.optString("state").takeUnless { it == "null" }
                }

                if (state == "exceeded" && previousState != "exceeded" &&
                    ClimateAlertSchedule.isActive(applicationContext, id, signature)) {
                    showNotification(applicationContext, location.optString("name"), alert)
                }

                states.put(
                    id,
                    JSONObject()
                        .put("signature", signature)
                        .put("state", state)
                        .put("evaluatedAt", now),
                )
            }
        }

        ClimateAlertSchedule.saveAlertStates(applicationContext, states)

        return if (hadRequestFailure) Result.retry() else Result.success()
    }

    private fun fetchCurrentWeather(endpoint: String, latitude: Double, longitude: Double): Map<String, Double?>? {
        if (!latitude.isFinite() || !longitude.isFinite() || latitude !in -90.0..90.0 || longitude !in -180.0..180.0) {
            return null
        }

        val uri = Uri.parse(endpoint).buildUpon()
            .appendQueryParameter("latitude", latitude.toString())
            .appendQueryParameter("longitude", longitude.toString())
            .appendQueryParameter("current", METRICS.joinToString(","))
            .appendQueryParameter("timezone", "auto")
            .appendQueryParameter("timeformat", "unixtime")
            .build()
        val connection = URL(uri.toString()).openConnection() as HttpURLConnection

        return try {
            connection.connectTimeout = 5_000
            connection.readTimeout = 8_000
            connection.requestMethod = "GET"

            if (connection.responseCode !in 200..299) {
                return null
            }

            val current = connection.inputStream.bufferedReader().use { reader ->
                JSONObject(reader.readText()).optJSONObject("current")
            } ?: return null

            METRICS.associateWith { metric ->
                val value = current.optDouble(metric, Double.NaN)
                value.takeIf { it.isFinite() }
            }
        } catch (_: Exception) {
            null
        } finally {
            connection.disconnect()
        }
    }

    private fun alertState(value: Double?, operator: String, threshold: Double): String {
        if (value == null || !threshold.isFinite()) {
            return "no_data"
        }

        val exceeded = when (operator) {
            "above" -> value > threshold
            "below" -> value < threshold
            else -> false
        }

        return if (exceeded) "exceeded" else "normal"
    }

    private fun showNotification(context: Context, locationName: String, alert: JSONObject) {
        if (!ClimateAlertSchedule.notificationAllowed(context)) {
            return
        }

        val locationIntent = context.packageManager.getLaunchIntentForPackage(context.packageName)
            ?.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP or Intent.FLAG_ACTIVITY_SINGLE_TOP)
        val contentIntent = locationIntent?.let {
            PendingIntent.getActivity(
                context,
                alert.optString("id").hashCode(),
                it,
                PendingIntent.FLAG_IMMUTABLE or PendingIntent.FLAG_UPDATE_CURRENT,
            )
        }
        val operator = if (alert.optString("operator") == "above") "superó" else "bajó de"
        val threshold = String.format(
            Locale.forLanguageTag("es-NI"),
            "%.1f",
            alert.optDouble("threshold"),
        )
        val message = "${alert.optString("label")} $operator $threshold ${alert.optString("unit")} en $locationName."
        val notification = NotificationCompat.Builder(context, "climate_alerts")
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .setContentTitle("Alerta climática")
            .setContentText(message)
            .setStyle(NotificationCompat.BigTextStyle().bigText(message))
            .setAutoCancel(true)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)

        contentIntent?.let(notification::setContentIntent)
        try {
            NotificationManagerCompat.from(context).notify(alert.optString("id").hashCode(), notification.build())
        } catch (_: SecurityException) {
            // The user can revoke permission between the check and notification delivery.
        }
    }

    private fun secureForecastUrl(value: String): String? {
        val candidate = value.ifBlank { "https://api.open-meteo.com/v1/forecast" }
        val uri = runCatching { Uri.parse(candidate) }.getOrNull() ?: return null

        return candidate.takeIf { uri.scheme == "https" && !uri.host.isNullOrBlank() }
    }

    companion object {
        private val METRICS = listOf(
            "temperature_2m",
            "relative_humidity_2m",
            "precipitation",
            "wind_speed_10m",
        )
    }
}

File: packages/agroclima/climate-notifications/resources/ios/ClimateNotificationFunctions.swift
import BackgroundTasks
import Foundation
import UIKit
import UserNotifications

enum ClimateNotificationFunctions {
    private static let permissionEvent = "App\\Events\\ClimateNotificationPermissionResult"

    final class CheckPermission: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let event = parameters["event"] as? String ?? permissionEvent
            let id = parameters["id"] as? String

            ClimateNotificationPermission.currentStatus { granted in
                ClimateNotificationPermission.dispatch(event: event, id: id, granted: granted)
            }

            return BridgeResponse.success(data: [:])
        }
    }

    final class RequestPermission: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            let event = parameters["event"] as? String ?? permissionEvent
            let id = parameters["id"] as? String

            UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound]) { granted, _ in
                ClimateNotificationPermission.dispatch(event: event, id: id, granted: granted)
            }

            return BridgeResponse.success(data: [:])
        }
    }

    final class SyncSchedule: BridgeFunction {
        func execute(parameters: [String: Any]) throws -> [String: Any] {
            guard JSONSerialization.isValidJSONObject(parameters),
                  let data = try? JSONSerialization.data(withJSONObject: parameters),
                  let schedule = String(data: data, encoding: .utf8) else {
                return BridgeResponse.error(code: "INVALID_SCHEDULE", message: "Invalid climate alert schedule")
            }

            ClimateAlertBackground.sync(schedule: schedule)

            return BridgeResponse.success(data: [:])
        }
    }

    static func initialize() {
        ClimateAlertBackground.initialize()
    }
}

private enum ClimateNotificationPermission {
    static func currentStatus(completion: @escaping (Bool) -> Void) {
        UNUserNotificationCenter.current().getNotificationSettings { settings in
            let allowed = settings.authorizationStatus == .authorized
                || settings.authorizationStatus == .provisional
                || settings.authorizationStatus == .ephemeral
            completion(allowed)
        }
    }

    static func dispatch(event: String, id: String?, granted: Bool) {
        var payload: [String: Any] = ["granted": granted]
        if let id {
            payload["id"] = id
        }

        DispatchQueue.main.async {
            LaravelBridge.shared.send?(event, payload)
        }
    }
}

private enum ClimateAlertBackground {
    private static let taskIdentifier = "dev.donmanuel.elver.climate-alert-refresh"
    private static let scheduleKey = "agroclima.climate-alerts.schedule"
    private static let alertStatesKey = "agroclima.climate-alerts.states"
    private static let defaultForecastUrl = "https://api.open-meteo.com/v1/forecast"
    private static let refreshInterval: TimeInterval = 60 * 60
    private static let metrics = [
        "temperature_2m",
        "relative_humidity_2m",
        "precipitation",
        "wind_speed_10m",
    ]
    private static let registered = BGTaskScheduler.shared.register(
        forTaskWithIdentifier: taskIdentifier,
        using: nil
    ) { task in
        guard let refreshTask = task as? BGAppRefreshTask else {
            task.setTaskCompleted(success: false)
            return
        }

        perform(refreshTask)
    }

    static func initialize() {
        guard registered else {
            return
        }

        scheduleStoredTask()
    }

    static func sync(schedule: String) {
        UserDefaults.standard.set(schedule, forKey: scheduleKey)

        guard let object = try? JSONSerialization.jsonObject(with: Data(schedule.utf8)) as? [String: Any],
              hasAlerts(in: object) else {
            cancel()
            UserDefaults.standard.set([String: [String: Any]](), forKey: alertStatesKey)
            return
        }

        reconcileStates(with: object)
        ClimateNotificationPermission.currentStatus { granted in
            if granted {
                scheduleNext()
            } else {
                cancel()
            }
        }
    }

    private static func perform(_ task: BGAppRefreshTask) {
        guard let schedule = storedSchedule(), hasAlerts(in: schedule) else {
            cancel()
            task.setTaskCompleted(success: true)
            return
        }

        scheduleNext()

        var operation: Task<Void, Never>?
        task.expirationHandler = {
            operation?.cancel()
        }

        operation = Task {
            let granted = await notificationsAllowed()
            let success = granted ? await refresh(schedule: schedule) : true

            if !granted {
                cancel()
            }

            task.setTaskCompleted(success: success)
        }
    }

    private static func refresh(schedule: [String: Any]) async -> Bool {
        guard let locations = schedule["locations"] as? [[String: Any]] else {
            return false
        }

        let forecastUrl = secureForecastUrl(schedule["forecastUrl"] as? String)
        var states = UserDefaults.standard.dictionary(forKey: alertStatesKey) as? [String: [String: Any]] ?? [:]
        var succeeded = true

        for location in locations {
            if Task.isCancelled {
                return false
            }

            guard let latitude = location["latitude"] as? Double,
                  let longitude = location["longitude"] as? Double,
                  let alerts = location["alerts"] as? [[String: Any]],
                  let current = await fetchCurrentWeather(
                    endpoint: forecastUrl,
                    latitude: latitude,
                    longitude: longitude
                  ) else {
                succeeded = false
                continue
            }

            let evaluatedAt = Date().timeIntervalSince1970 * 1_000

            for alert in alerts {
                guard let id = alert["id"] as? String,
                      let signature = alert["signature"] as? String,
                      let metric = alert["metric"] as? String,
                      let operatorValue = alert["operator"] as? String,
                      let threshold = alert["threshold"] as? Double else {
                    continue
                }

                guard isActive(id: id, signature: signature) else {
                    continue
                }

                let state = alertState(
                    current[metric],
                    comparison: operatorValue,
                    threshold: threshold
                )
                let saved = states[id]
                let previousState = saved?["signature"] as? String == signature
                    ? saved?["state"] as? String
                    : alert["lastState"] as? String

                if state == "exceeded", previousState != "exceeded",
                   isActive(id: id, signature: signature) {
                    await showNotification(location: location, alert: alert)
                }

                states[id] = [
                    "signature": signature,
                    "state": state,
                    "evaluatedAt": evaluatedAt,
                ]
            }
        }

        let activeAlerts = activeAlertSignatures()
        let activeStates = states.filter { id, state in
            activeAlerts[id] == state["signature"] as? String
        }
        UserDefaults.standard.set(activeStates, forKey: alertStatesKey)
        return succeeded
    }

    private static func fetchCurrentWeather(
        endpoint: URL,
        latitude: Double,
        longitude: Double
    ) async -> [String: Double]? {
        guard (-90...90).contains(latitude), (-180...180).contains(longitude),
              var components = URLComponents(url: endpoint, resolvingAgainstBaseURL: false) else {
            return nil
        }

        components.queryItems = [
            URLQueryItem(name: "latitude", value: String(latitude)),
            URLQueryItem(name: "longitude", value: String(longitude)),
            URLQueryItem(name: "current", value: metrics.joined(separator: ",")),
            URLQueryItem(name: "timezone", value: "auto"),
            URLQueryItem(name: "timeformat", value: "unixtime"),
        ]

        guard let url = components.url else {
            return nil
        }

        var request = URLRequest(url: url)
        request.timeoutInterval = 10

        do {
            let (data, response) = try await URLSession.shared.data(for: request)
            guard let response = response as? HTTPURLResponse,
                  (200...299).contains(response.statusCode),
                  let payload = try JSONSerialization.jsonObject(with: data) as? [String: Any],
                  let current = payload["current"] as? [String: Any] else {
                return nil
            }

            var values: [String: Double] = [:]

            for metric in metrics {
                if let value = current[metric] as? Double, value.isFinite {
                    values[metric] = value
                }
            }

            return values
        } catch {
            return nil
        }
    }

    private static func alertState(_ value: Double?, comparison: String, threshold: Double) -> String {
        guard let value, value.isFinite, threshold.isFinite else {
            return "no_data"
        }

        let exceeded = comparison == "above" ? value > threshold : comparison == "below" ? value < threshold : false
        return exceeded ? "exceeded" : "normal"
    }

    private static func showNotification(location: [String: Any], alert: [String: Any]) async {
        let content = UNMutableNotificationContent()
        let name = location["name"] as? String ?? "tu ubicación"
        let label = alert["label"] as? String ?? "Una variable climática"
        let unit = alert["unit"] as? String ?? ""
        let operatorPhrase = alert["operator"] as? String == "above" ? "superó" : "bajó de"
        let threshold = alert["threshold"] as? Double ?? 0

        content.title = "Alerta climática"
        content.body = String(
            format: "%@ %@ %.1f %@ en %@.",
            locale: Locale(identifier: "es_NI"),
            label,
            operatorPhrase,
            threshold,
            unit,
            name
        )
        content.sound = .default

        let request = UNNotificationRequest(
            identifier: "climate-alert-\(alert["id"] as? String ?? UUID().uuidString)",
            content: content,
            trigger: nil
        )
        do {
            try await UNUserNotificationCenter.current().add(request)
        } catch {
            NSLog("Unable to post climate alert notification: %@", error.localizedDescription)
        }
    }

    private static func reconcileStates(with schedule: [String: Any]) {
        guard let locations = schedule["locations"] as? [[String: Any]] else {
            return
        }

        let previous = UserDefaults.standard.dictionary(forKey: alertStatesKey) as? [String: [String: Any]] ?? [:]
        var next: [String: [String: Any]] = [:]

        for location in locations {
            guard let alerts = location["alerts"] as? [[String: Any]] else {
                continue
            }

            for alert in alerts {
                guard let id = alert["id"] as? String,
                      let signature = alert["signature"] as? String else {
                    continue
                }

                let timestamp = (alert["lastEvaluatedAt"] as? NSNumber)?.doubleValue ?? 0
                let saved = previous[id]
                if saved?["signature"] as? String != signature
                    || timestamp > (saved?["evaluatedAt"] as? Double ?? 0) {
                    next[id] = [
                        "signature": signature,
                        "state": alert["lastState"] as? String ?? "",
                        "evaluatedAt": timestamp,
                    ]
                } else if let saved {
                    next[id] = saved
                }
            }
        }

        UserDefaults.standard.set(next, forKey: alertStatesKey)
    }

    private static func hasAlerts(in schedule: [String: Any]) -> Bool {
        (schedule["locations"] as? [[String: Any]])?.contains {
            !($0["alerts"] as? [[String: Any]] ?? []).isEmpty
        } ?? false
    }

    private static func isActive(id: String, signature: String) -> Bool {
        activeAlertSignatures()[id] == signature
    }

    private static func activeAlertSignatures() -> [String: String] {
        guard let schedule = storedSchedule(),
              let locations = schedule["locations"] as? [[String: Any]] else {
            return [:]
        }

        var signatures: [String: String] = [:]
        for location in locations {
            for alert in (location["alerts"] as? [[String: Any]] ?? []) {
                guard let id = alert["id"] as? String,
                      let signature = alert["signature"] as? String else {
                    continue
                }
                signatures[id] = signature
            }
        }

        return signatures
    }

    private static func storedSchedule() -> [String: Any]? {
        guard let raw = UserDefaults.standard.string(forKey: scheduleKey),
              let data = raw.data(using: .utf8) else {
            return nil
        }

        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }

    private static func scheduleStoredTask() {
        guard let schedule = storedSchedule(), hasAlerts(in: schedule) else {
            return
        }

        ClimateNotificationPermission.currentStatus { granted in
            if granted {
                scheduleNext()
            }
        }
    }

    private static func scheduleNext() {
        let request = BGAppRefreshTaskRequest(identifier: taskIdentifier)
        request.earliestBeginDate = Date(timeIntervalSinceNow: refreshInterval)

        do {
            try BGTaskScheduler.shared.submit(request)
        } catch {
            NSLog("Unable to schedule climate alert refresh: %@", error.localizedDescription)
        }
    }

    private static func cancel() {
        BGTaskScheduler.shared.cancel(taskRequestWithIdentifier: taskIdentifier)
    }

    private static func notificationsAllowed() async -> Bool {
        await withCheckedContinuation { continuation in
            ClimateNotificationPermission.currentStatus { granted in
                continuation.resume(returning: granted)
            }
        }
    }

    private static func secureForecastUrl(_ value: String?) -> URL {
        let candidate = URL(string: value ?? defaultForecastUrl)
        if let candidate, candidate.scheme == "https", candidate.host != nil {
            return candidate
        }

        return URL(string: defaultForecastUrl)!
    }
}
